File: Python/ilinkai_weixin/cdn/cdn_client.py
"""CDN客户端服务"""
import base64
import logging
from urllib.request import Request, urlopen
from urllib.error import HTTPError

from .aes_ecb_crypto import AesEcbCrypto
from .cdn_url_builder import CdnUrlBuilder

logger = logging.getLogger(__name__)
_MAX_RETRIES = 3


class CdnUploadException(Exception):
    def __init__(self, message: str, status_code: int = 0):
        super().__init__(message)
        self.status_code = status_code


class CdnDownloadException(Exception):
    def __init__(self, message: str, status_code: int = 0):
        super().__init__(message)
        self.status_code = status_code


class CdnClient:
    """CDN客户端服务 - 处理文件的上传和下载"""

    def upload_buffer(self, buffer: bytes, upload_param: str, file_key: str,
                      cdn_base_url: str, aes_key: bytes, label: str = "upload") -> str:
        """上传缓冲区到CDN（带AES-128-ECB加密）"""
        ciphertext = AesEcbCrypto.encrypt(buffer, aes_key)
        cdn_url = CdnUrlBuilder.build_upload_url(upload_param, file_key, cdn_base_url)
        logger.debug("%s: CDN POST url=%s ciphertextSize=%d", label, _redact_url(cdn_url), len(ciphertext))

        last_error = None
        for attempt in range(1, _MAX_RETRIES + 1):
            try:
                req = Request(cdn_url, data=ciphertext, method='POST')
                req.add_header('Content-Type', 'application/octet-stream')
                with urlopen(req) as resp:
                    status = resp.status
                    if status < 200 or status >= 300:
                        raise CdnUploadException(f"CDN upload server error: status {status}", status)
                    download_param = resp.headers.get('x-encrypted-param')
                    if not download_param:
                        raise CdnUploadException("CDN upload response missing x-encrypted-param header")
                    logger.debug("%s: CDN upload success attempt=%d", label, attempt)
                    return download_param
            except HTTPError as e:
                if 400 <= e.code < 500:
                    err_msg = e.headers.get('x-error-message') or (e.read().decode('utf-8') if e.fp else '')
                    raise CdnUploadException(f"CDN upload client error {e.code}: {err_msg}", e.code)
                err_msg = e.headers.get('x-error-message') or f"status {e.code}"
                raise CdnUploadException(f"CDN upload server error: {err_msg}", e.code)
            except CdnUploadException:
                raise
            except Exception as e:
                last_error = e
                if attempt < _MAX_RETRIES:
                    logger.error("%s: attempt %d failed, retrying... err=%s", label, attempt, e)

        raise CdnUploadException(f"CDN upload failed after {_MAX_RETRIES} attempts: {last_error}")

    def download_and_decrypt(self, encrypted_query_param: str, aes_key_base64: str,
                              cdn_base_url: str, label: str = "download") -> bytes:
        """从CDN下载并解密数据"""
        key = _parse_aes_key(aes_key_base64, label)
        url = CdnUrlBuilder.build_download_url(encrypted_query_param, cdn_base_url)
        logger.debug("%s: fetching url=%s", label, _redact_url(url))
        encrypted = _fetch_cdn_bytes(url, label)
        logger.debug("%s: downloaded %d bytes, decrypting", label, len(encrypted))
        decrypted = AesEcbCrypto.decrypt(encrypted, key)
        logger.debug("%s: decrypted %d bytes", label, len(decrypted))
        return decrypted

    def download_plain(self, encrypted_query_param: str, cdn_base_url: str,
                        label: str = "download") -> bytes:
        """从CDN下载原始数据（不解密）"""
        url = CdnUrlBuilder.build_download_url(encrypted_query_param, cdn_base_url)
        logger.debug("%s: fetching url=%s", label, _redact_url(url))
        return _fetch_cdn_bytes(url, label)


def _fetch_cdn_bytes(url: str, label: str) -> bytes:
    try:
        with urlopen(url) as resp:
            if resp.status < 200 or resp.status >= 300:
                body = resp.read().decode('utf-8', errors='replace')
                raise CdnDownloadException(f"{label}: CDN download {resp.status} body={body}", resp.status)
            return resp.read()
    except CdnDownloadException:
        raise
    except HTTPError as e:
        body = e.read().decode('utf-8', errors='replace') if e.fp else ''
        raise CdnDownloadException(f"{label}: CDN download {e.code} body={body}", e.code)
    except Exception as e:
        raise CdnDownloadException(f"{label}: fetch network error: {e}")


def _parse_aes_key(aes_key_base64: str, label: str) -> bytes:
    decoded = base64.b64decode(aes_key_base64)
    if len(decoded) == 16:
        return decoded
    if len(decoded) == 32:
        hex_string = decoded.decode('ascii', errors='replace')
        if all(c in '0123456789abcdefABCDEF' for c in hex_string):
            return bytes.fromhex(hex_string)
    raise ValueError(f"{label}: aes_key must decode to 16 raw bytes or 32-char hex string, got {len(decoded)} bytes")


def _redact_url(url: str) -> str:
    idx = url.find('?')
    return url[:idx] + '?[REDACTED]' if idx >= 0 else url

File: Java/src/main/java/com/ilinkai/weixin/cdn/CdnClient.java
package com.ilinkai.weixin.cdn;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

/** CDN客户端服务 - 处理文件的上传和下载 */
public class CdnClient {
    private static final Logger logger = LoggerFactory.getLogger(CdnClient.class);
    private static final int MAX_RETRIES = 3;

    /** 上传缓冲区到CDN（带AES-128-ECB加密） */
    public String uploadBuffer(byte[] buffer, String uploadParam, String fileKey,
                               String cdnBaseUrl, byte[] aesKey, String label) throws CdnException {
        String cdnUrl = CdnUrlBuilder.buildUploadUrl(uploadParam, fileKey, cdnBaseUrl);
        return uploadBufferCore(buffer, cdnUrl, aesKey, label);
    }

    /** 上传缓冲区到CDN（使用完整URL） */
    public String uploadBufferByUrl(byte[] buffer, String uploadFullUrl,
                                     byte[] aesKey, String label) throws CdnException {
        return uploadBufferCore(buffer, uploadFullUrl, aesKey, label);
    }

    private String uploadBufferCore(byte[] buffer, String cdnUrl, byte[] aesKey, String label) throws CdnException {
        byte[] ciphertext;
        try {
            ciphertext = AesEcbCrypto.encrypt(buffer, aesKey);
        } catch (Exception e) {
            throw new CdnException("AES encryption failed: " + e.getMessage());
        }
        logger.debug("{}: CDN POST url={} ciphertextSize={}", label, redactUrl(cdnUrl), ciphertext.length);

        Exception lastError = null;
        for (int attempt = 1; attempt <= MAX_RETRIES; attempt++) {
            try {
                HttpURLConnection conn = (HttpURLConnection) new URL(cdnUrl).openConnection();
                conn.setRequestMethod("POST");
                conn.setDoOutput(true);
                conn.setRequestProperty("Content-Type", "application/octet-stream");

                try (OutputStream os = conn.getOutputStream()) {
                    os.write(ciphertext);
                }

                int status = conn.getResponseCode();

                if (status >= 400 && status < 500) {
                    String errMsg = conn.getHeaderField("x-error-message");
                    if (errMsg == null) errMsg = readStream(conn.getErrorStream());
                    throw new CdnException("CDN upload client error " + status + ": " + errMsg, status);
                }

                if (status < 200 || status >= 300) {
                    String errMsg = conn.getHeaderField("x-error-message");
                    if (errMsg == null) errMsg = "status " + status;
                    throw new CdnException("CDN upload server error: " + errMsg, status);
                }

                String downloadParam = conn.getHeaderField("x-encrypted-param");
                if (downloadParam == null || downloadParam.isEmpty()) {
                    throw new CdnException("CDN upload response missing x-encrypted-param header");
                }

                logger.debug("{}: CDN upload success attempt={}", label, attempt);
                return downloadParam;

            } catch (CdnException e) {
                throw e;
            } catch (Exception e) {
                lastError = e;
                if (attempt < MAX_RETRIES) {
                    logger.error("{}: attempt {} failed, retrying... err={}", label, attempt, e.getMessage());
                }
            }
        }
        throw new CdnException("CDN upload failed after " + MAX_RETRIES + " attempts: " +
                (lastError != null ? lastError.getMessage() : "unknown"));
    }

    /** 从CDN下载并解密数据 */
    public byte[] downloadAndDecrypt(String encryptedQueryParam, String aesKeyBase64,
                                      String cdnBaseUrl, String label) throws CdnException {
        byte[] key = parseAesKey(aesKeyBase64, label);
        String url = CdnUrlBuilder.buildDownloadUrl(encryptedQueryParam, cdnBaseUrl);
        logger.debug("{}: fetching url={}", label, redactUrl(url));

        byte[] encrypted = fetchCdnBytes(url, label);
        logger.debug("{}: downloaded {} bytes, decrypting", label, encrypted.length);

        try {
            byte[] decrypted = AesEcbCrypto.decrypt(encrypted, key);
            logger.debug("{}: decrypted {} bytes", label, decrypted.length);
            return decrypted;
        } catch (Exception e) {
            throw new CdnException(label + ": AES decryption failed: " + e.getMessage());
        }
    }

    /** 从CDN下载原始数据（不解密） */
    public byte[] downloadPlain(String encryptedQueryParam, String cdnBaseUrl, String label) throws CdnException {
        String url = CdnUrlBuilder.buildDownloadUrl(encryptedQueryParam, cdnBaseUrl);
        logger.debug("{}: fetching url={}", label, redactUrl(url));
        return fetchCdnBytes(url, label);
    }

    private byte[] fetchCdnBytes(String url, String label) throws CdnException {
        try {
            HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
            conn.setRequestMethod("GET");
            int status = conn.getResponseCode();

            if (status < 200 || status >= 300) {
                String body = readStream(conn.getErrorStream());
                throw new CdnException(label + ": CDN download " + status + " " + conn.getResponseMessage() + " body=" + body, status);
            }

            return readBytes(conn.getInputStream());
        } catch (CdnException e) {
            throw e;
        } catch (Exception e) {
            throw new CdnException(label + ": fetch network error: " + e.getMessage());
        }
    }

    private static byte[] parseAesKey(String aesKeyBase64, String label) throws CdnException {
        byte[] decoded = Base64.getDecoder().decode(aesKeyBase64);
        if (decoded.length == 16) return decoded;
        if (decoded.length == 32) {
            String hexString = new String(decoded, StandardCharsets.US_ASCII);
            if (isHexString(hexString)) {
                return hexStringToBytes(hexString);
            }
        }
        throw new CdnException(label + ": aes_key must decode to 16 raw bytes or 32-char hex string, got " + decoded.length + " bytes");
    }

    private static boolean isHexString(String s) {
        if (s.length() != 32) return false;
        for (char c : s.toCharArray()) {
            if ("0123456789abcdefABCDEF".indexOf(c) < 0) return false;
        }
        return true;
    }

    static byte[] hexStringToBytes(String hex) {
        byte[] result = new byte[hex.length() / 2];
        for (int i = 0; i < result.length; i++) {
            result[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return result;
    }

    static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) sb.append(String.format("%02x", b));
        return sb.toString();
    }

    private static String readStream(InputStream is) throws IOException {
        if (is == null) return "";
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] buf = new byte[4096];
        int n;
        while ((n = is.read(buf)) != -1) baos.write(buf, 0, n);
        return baos.toString(StandardCharsets.UTF_8.name());
    }

    private static byte[] readBytes(InputStream is) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] buf = new byte[8192];
        int n;
        while ((n = is.read(buf)) != -1) baos.write(buf, 0, n);
        return baos.toByteArray();
    }

    private static String redactUrl(String url) {
        int idx = url.indexOf('?');
        return idx >= 0 ? url.substring(0, idx) + "?[REDACTED]" : url;
    }
}
